packages/gatsby-cli/src/create-cli.js

```diff
@@ -71,8 +85,33 @@ function buildLocalCommands(cli, isLocalSite) {
           `There was a problem loading the local ${command} command. Gatsby may not be installed in your site's "node_modules" directory. Perhaps you need to run "npm install"? You might need to delete your "package-lock.json" as well.`
         )
 
-      report.verbose(`loading local command from: ${cmdPath}`)
-      return require(cmdPath)
+      if (command === `develop`) {
+        return args => {
+          const script = createControllableScript(`
+            const cmd = require("${cmdPath}");
+            const args = ${JSON.stringify(args)};
+            cmd(args);
+          `)
+
+          script.start()
+
+          chokidar
+            .watch([rootFile(`gatsby-config.js`), rootFile(`gatsby-node.js`)])
+            .on(`change`, filePath => {
+              const activity = report.activityTimer(
+                `${path.basename(filePath)} changed, restarting gatsby develop`
+              )
+              activity.start()
+              script.stop(() => {
+                activity.end()
+                script.start()
+              })
+            })
+        }
+      } else {
+        report.verbose(`loading local command from: ${cmdPath}`)
+        return require(cmdPath)
+      }
```

Comment:
Couple thoughts.

1. Could this not live in `gatsby/commands/develop` ?
1. If not, we should move this code out into it's own module. I feel like this muddy's up this file.

I'm mostly wondering if we do it in develop if we can do it a bit cleaner than stopping and spawning a new node instance. This might have negative consequences down the road. Like when we start building a more interactive develop CLI. I expect that would create some weird side-effects.

Reply:
> Could this not live in gatsby/commands/develop ?

Sure, I could change `develop` to fork the process and run a script in the `gatsby` package, happy to do it that way if you prefer!

> I'm mostly wondering if we do it in develop if we can do it a bit cleaner than stopping and spawning a new node instance. This might have negative consequences down the road. Like when we start building a more interactive develop CLI. I expect that would create some weird side-effects.

We have to do a full restart with all the bootstrapping so we can be sure all config changes are applied, right? I think the interactive part should not really be affected by this, given stdio is just inherited through? (I'm not a Node expert so that might be wrong 😅)